syntax = "proto3";

package transport_catalogue_serialize;

message ObjectName {
  string name = 1;
  uint32 id = 2;
}

message RoadDistances {
  uint32 name_stop_to = 1;
  int32 distances = 2;
}

message Stop {
  uint32 name = 1;
  double latitude = 2;
  double longitude = 3;
  repeated RoadDistances road_distances = 4;
}

message Bus {
  uint32 name = 1;
  repeated int32 stops = 2;
  bool is_roundtrip = 3;
}

message TransportCatalogue {
  repeated ObjectName name_repository = 1;
  repeated Stop stops = 2;
  repeated Bus buses = 3;
}

message Rgb {
  uint32 red = 1;
  uint32 green = 2;
  uint32 blue = 3;
}

message Rgba {
  uint32 red = 1;
  uint32 green = 2;
  uint32 blue = 3;
  double opacity = 4;
}

message ColorStr {
  string color = 1;
}

message Color {
  ColorStr collorstr = 1;
  Rgb rgb = 2;
  Rgba rgba = 3;
}

message VisualizationSettings {
  double width = 1;
  double height = 2;
  double padding = 3;
  double line_width = 4;
  double stop_radius = 5;
  double underlayer_width = 6;
  int32 bus_label_font_size = 7;
  int32 stop_label_font_size = 8;
  repeated double bus_label_offset = 9;
  repeated double stop_label_offset = 10;
  Color underlayer_color = 11;
  repeated Color color_palette = 12;
}

message RouteSettings {
  double bus_velocity = 1;
  double bus_wait_time = 2;
}

message SerializationSetting {
    TransportCatalogue transport_catalogue = 1;
    VisualizationSettings vs = 2;
    RouteSettings rs = 3;
}
